### PR TITLE
Restore but deprecate Eff and legacy handlers.

### PR DESCRIPTION
**Pull request by patrickt**
Because the next release contains significant API churn, we've decided
to make the path easier by restoring but deprecating the legacy idioms
for defining effects. This restores the `handleSum` family of
functions while informing the user that they will be removed in
upcoming releases, as well as defining `Eff m` as a synonym for `m`
and marking it as deprecated.

### CODE REVIEW

File: src/Control/Effect.hs [hunk 1]
```diff
@@ -19,3 +20,6 @@ import Control.Effect.State     as X (State, StateC)
 import Control.Effect.Sum       as X ((:+:), Member, send)
 import Control.Effect.Trace     as X (Trace, TraceByPrintingC, TraceByIgnoringC, TraceByReturningC)
 import Control.Effect.Writer    as X (Writer, WriterC)
+
+type Eff m = m
```
Comment (patrickt) [inline src/Control/Effect.hs]: There’s no `Internal` module anymore, so I’m open to suggestions as to where to place this until we kick it out in 0.4.0.

File: src/Control/Effect/Carrier.hs [hunk 2]
```diff
@@ -40,6 +46,11 @@ class (HFunctor sig, Monad m) => Carrier sig m | m -> sig where
   -- | Construct a value in the carrier for an effect signature (typically a sum of a handled effect and any remaining effects).
   eff :: sig m (m a) -> m a
 
+  -- | Construct a value in the carrier for an effect signature (typically a sum of a handled effect and any remaining effects).
+  ret :: a -> m a
+  ret = pure
```
Comment (patrickt) [inline src/Control/Effect/Carrier.hs]: Made this default to `pure`.

File: ChangeLog.md [hunk 3]
```diff
@@ -1,22 +1,31 @@
...
 - Replaces `AltC` with a new carrier, `NonDetC`, based on Ralf Hinze’s work in _[Deriving Backtracking Monad Transformers](https://www.cs.ox.ac.uk/ralf.hinze/publications/#P12)_.
   This is a backwards-incompatible change. `AltC` was equivalent to the `ListT` monad transformer, and had the same well-known limitation to commutative monads. Therefore, the elimination of `Eff` required a more durable approach.
 - Removes `Branch`.
   This is a backwards-incompatible change, but was necessitated by the difficulty of implementing correct `Applicative` & `Monad` instances for carriers which used it. Carriers which were employing `Branch` internally should be reimplemented using `NonDetC` or a similar approach; see `CutC` and `CullC` for examples.
+- Renames `Control.Effect.Void`, `Void`, and `VoidC` to `Control.Effect.Pure`, `Pure`, and `PureC` respectively.
+  This is a backwards-incompatible change for code mentioning `VoidC`; it should be updated to reference `PureC` instead.
+
+## Deprecations
```
Comment (robrix) [inline ChangeLog.md]: I particularly dig this organization of the changelog 👍

File: ChangeLog.md [hunk 4]
```diff
@@ -1,22 +1,31 @@
...
+   Carriers can be composed from other carriers and `eff` defined with `handleCoercible`; and other definitions can use `handlePure` & `handle` directly.
+
+All deprecated APIs will be removed in the next release.
+
+## Other changes
+
+- Adds `Monad` instances for all carrier types.
 - Rewrites `CutC` using an approach related to `NonDetC`, with the addition of a continuation to distinguish `empty` from `cutfail`.
 - Rewrites `CullC` using `ListC` and `ReaderC`.
 - Moves `OnceC` from `Control.Effect.NonDet` to `Control.Effect.Cull` to avoid cyclic dependencies.
```
Comment (robrix) [inline ChangeLog.md]: These three could be considered backwards-incompatible for any code that depended on the structure of these effects; but I think the likelihood of that is pretty low.

File: src/Control/Effect.hs (hunk 1 above)
Comment (robrix) [reply src/Control/Effect.hs]: This is a fine place for it 👍 Further, I think this is a really good definition to use for it in the interim.

File: src/Control/Effect/Carrier.hs (hunk 2 above)
Comment (robrix) [reply src/Control/Effect/Carrier.hs]: This is an excellent default 👍

File: src/Control/Effect/Carrier.hs [hunk 5]
```diff
@@ -52,3 +63,31 @@ handlePure handler = hmap handler . fmap' handler
...
+
+-- | Thread a carrier producing values in a 'Traversable' 'Monad' (e.g. '[]') through an 'Effect'.
+handleTraversable :: (Effect sig, Applicative g, Monad m, Traversable m) => (forall x . f x -> g (m x)) -> sig f (f a) -> sig g (g (m a))
+handleTraversable run = handle (pure ()) (fmap join . traverse run)
+{-# INLINE handleTraversable #-}
+
+-- | A backwards-compatibility shim, equivalent to 'id'.
+interpret :: carrier a -> carrier a
+interpret = id
+{-# DEPRECATED interpret "Not necessary with monadic carriers; remove or replace with 'id'." #-}
```
Comment (robrix) [inline src/Control/Effect/Carrier.hs]: Love the thought you’ve put into definitions like these and the deprecation messages 👍

File: src/Control/Effect/Carrier.hs [hunk 6]
```diff
@@ -52,3 +63,31 @@ handlePure handler = hmap handler . fmap' handler
 handleCoercible :: (HFunctor sig, Coercible f g) => sig f (f a) -> sig g (g a)
 handleCoercible = handlePure coerce
 {-# INLINE handleCoercible #-}
+
+{-# DEPRECATED handleReader, handleState, handleEither, handleTraversable
+  "Compose carrier types from other carriers and define 'eff' with handleCoercible instead" #-}
```
Comment (robrix) [inline src/Control/Effect/Carrier.hs]: May also note that `handle`/`handlePure` can be used for non-compositional carriers.

Comment (robrix) [reply src/Control/Effect/Carrier.hs]: @patrickt: I think this got missed; what do you think about mentioning `handle`/`handlePure` here?

**Pull request merged**